src/treemodel/selection.js

```diff
@@ -212,54 +274,230 @@ export default class Selection {
 	 */
 	setAttributesTo( attrs ) {
 		this._attrs = utils.toMap( attrs );
+		this._setStoredAttributesTo( this._attrs );
 	}
 
 	/**
-	 * Removes an attribute with given key from the selection.
+	 * Converts given range to {@link core.treeModel.LiveRange} and adds it to internal ranges array. Throws an error
+	 * if given range is intersecting with any range that is already stored in this selection.
+	 *
+	 * @private
+	 * @param {core.treeModel.Range} range Range to add.
+	 */
+	_pushRange( range ) {
+		for ( let i = 0; i < this._ranges.length ; i++ ) {
+			if ( range.isIntersecting( this._ranges[ i ] ) ) {
+				/**
+				 * Trying to add a range that intersects with another range from selection.
+				 *
+				 * @error selection-range-intersects
+				 * @param {core.treeModel.Range} addedRange Range that was added to the selection.
+				 * @param {core.treeModel.Range} intersectingRange Range from selection that intersects with `addedRange`.
+				 */
+				throw new CKEditorError(
+					'selection-range-intersects: Trying to add a range that intersects with another range from selection.',
+					{ addedRange: range, intersectingRange: this._ranges[ i ] }
+				);
+			}
+		}
+
+		this._ranges.push( LiveRange.createFromRange( range ) );
+	}
+
+	/**
+	 * Iterates through all attributes stored in current selection's parent.
 	 *
+	 * @returns {Iterable.<*>}
+	 */
+	*_getStoredAttributes() {
```

Comment:
It's been a bit confusing for me that in one place you have `set/get/removeAttribute()` and next to it you have `_storeAttribute()` and the rest. We need to find a more distinguishable name for the "hoist attributes to the block" functionality.

Some ideas:

* `hoistAttribute()`, `removeHoistedAttribute()`,
* `setBlockAttribute()`, `removeBlockAttribute()`,
* `preserveAttribute()`